File: golang-http-server/internal/middleware/error_handler.go
package middleware

import (
	"encoding/json"
	"net/http"

	"golang-http-server/internal/core/errors"
	"golang-http-server/internal/dto"
)

func ErrorHandler(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		defer func() {
			if err := recover(); err != nil {
				w.Header().Set("Content-Type", "application/json")
				w.WriteHeader(http.StatusInternalServerError)
				json.NewEncoder(w).Encode(dto.ErrorResponse{
					Error: "internal server error",
				})
			}
		}()

		next.ServeHTTP(w, r)
	})
}

func WriteError(w http.ResponseWriter, err error) {
	w.Header().Set("Content-Type", "application/json")

	var appErr *errors.AppError
	if e, ok := err.(*errors.AppError); ok {
		appErr = e
	} else {
		appErr = &errors.AppError{
			Message:      "internal server error",
			StatusCode:   500,
			IsOperational: false,
		}
	}

	w.WriteHeader(appErr.StatusCode)
	json.NewEncoder(w).Encode(dto.ErrorResponse{
		Error: appErr.Message,
	})
}

func WriteJSON(w http.ResponseWriter, status int, data any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(data)
}

File: hono-node/src/middleware/error-handler.ts
import type { MiddlewareHandler } from "hono";

import { AppError } from "@/core/errors/app-error";
import type { HonoEnv } from "@/core/hono-env";
import { resolveApiLogger } from "@/core/logger";

const STATUS_CODES: Record<
  number,
  400 | 401 | 403 | 404 | 409 | 422 | 429 | 500 | 501 | 503
> = {
  400: 400,
  401: 401,
  403: 403,
  404: 404,
  409: 409,
  422: 422,
  429: 429,
  500: 500,
  501: 501,
  503: 503,
};

export const errorHandler: MiddlewareHandler<HonoEnv> = async (c, next) => {
  const logger = resolveApiLogger();

  try {
    await next();
  } catch (err) {
    if (err instanceof AppError) {
      logger.warn(
        { statusCode: err.statusCode, message: err.message },
        "Operational error",
      );
      const status = STATUS_CODES[err.statusCode] ?? 500;
      return c.json({ error: err.message }, status);
    }

    logger.error({ err }, "Unexpected error");
    return c.json({ error: "Internal server error" }, 500);
  }
};
